mhkit/wave/contours.py

```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+    sample_u2 = sample_beta * np.sin(sample_alpha)
+
+
+    comp1_sample = stats.invgauss.ppf(
+        stats.norm.cdf(sample_u1, loc=0, scale=1),
+        mu=comp1['mu'], loc=0, scale=comp1['scale'])
+    mu_sample = mu_slope * comp1_sample + mu_intercept
+
+    # Calculate sigma values at each point on the circle
+    sigma_sample = np.polyval(sigma_polynomial_coeffcients, comp1_sample)
+
+    # Use calculated mu and sigma values to calculate C2 along the contour
+    comp2_sample = stats.norm.ppf(stats.norm.cdf(sample_u2, loc=0, scale=1),
+                                  loc=mu_sample, scale=sigma_sample)
+
+    # Sample transformation into Hs-T space
+    h_sample, t_sample = _princomp_inv(
+        comp1_sample, comp2_sample, coeff, shift)
+
+    return h_sample, t_sample, weight_points
+
+
+def samples_contour(t_samples, t_contour, hs_contour):
+    """Get Hs points along a specified environmental contour using
+    user-defined T values.
+
+    Parameters
+    ----------
+    t_samples : nparray
+        Points for sampling along return contour
+    t_contour : np.array
+        T values along contour
+    hs_contour : np.array
+        Hs values along contour
+
+    Returns
+    -------
+    hs_samples : nparray
+        points sampled along return contour
+    """
+    #finds minimum and maximum energy period values
+    amin = np.argmin(t_contour)
+    amax = np.argmax(t_contour)
+    aamin = np.min([amin, amax])
+    aamax = np.max([amin, amax])
+    #finds points along the contour
+    w1 = hs_contour[aamin:aamax]
+    w2 = np.concatenate((hs_contour[aamax:], hs_contour[:aamin]))
+    if (np.max(w1) > np.max(w2)):
+        x1 = t_contour[aamin:aamax]
+        y = hs_contour[aamin:aamax]
+    else:
+        x1 = np.concatenate((t_contour[aamax:], t_contour[:aamin]))
+        y1 = np.concatenate((hs_contour[aamax:], hs_contour[:aamin]))
+    #sorts data based on the max and min energy period values
+    ms = np.argsort(x1)
+    x = x1[ms]
+    y = y1[ms]
+    #interpolates the sorted data
+    si = interp.interp1d(x, y)
+    #finds the wave height based on the user specified energy period values
+    hs_samples = si(t_samples)
+
+    return hs_samples
+
+
+def _generate_sample_data(beta_lines, rho_zeroline, theta_zeroline,
+                          points_per_interval, contour_probs):
+    """ Calculate radius, angle, and weight for each sample point
+
+    Parameters
+    ----------
+    beta_lines: np.array
+        Array of mu fitting function parameters.
+    rho_zeroline: np.array
+        Array of radii
+    theta_zeroline: np.array
+    points_per_interval: np.array
+    contour_probs: np.array
+
+    Returns
+    -------
+    sample_alpha: np.array
+        Array of fitted sample angle values.
+    sample_beta: np.array
+        Array of fitted sample radius values.
+    weight_points: np.array
+        Array of weights for each point.
+    """
+    num_samples = (len(beta_lines) - 1) * points_per_interval
```

Comment:
Internal functions also need to have assert statements to check the incoming data. This allows for changes to function upstream of this to be quickly caught for and errors in data formatting this function may expect.

Reply:
Done w/ @hivanov-nrel latest commit.